### ISSUE TITLE
Standardize Analyzer Tests

### ISSUE THREAD

**Issue by tehsphinx** [exercism/automated-mentoring-support#16]
When it comes to analyzers producing comments I consider this a very good case for TDD. Tests for exalysis do this by taking real life solutions and defining which comment to add or not to add.

Should we generalize that accross analyzers or leave that up to the implementers?

I would vote for generalizing that so that contributors outside the specific analyzer team can easily add a new case to be covered by the analyzer.

Here my suggestion:
- Have a folder per exercise with one subfolder for each sample solution
- have a `cases` file in each exercise folder with a list of which solution folder should provide/not provide which comment. The format could be language specific (like the tests are) or language agnostic (e.g. JSON, YAML).
- location of actual test is left up to the analyzer team.

Folder structure:
```
- tests
   | - twofer
        | - solution1
        | - solution2
        | - solution3
        | cases.yml
   | - hamming
        | - solution1
        | - solution2
        | cases.yml
```

Here a sample of how exalysis does it currently: https://github.com/tehsphinx/exalysis/blob/master/track/twofer/twofer_test.go

**Comment by tehsphinx** [comment]
A sample on how the `cases.json` file could look like:

```json
[
  {
    "solution": "two-fer/solution1",
    "expected_comments": [
      "go.two_fer.plus_used",
      "go.two_fer.remove_stub"
    ],
    "not_expected_comments": [
      "go.two_fer.missing_comment",
      "go.two_fer.comment_format",
      "go.two_fer.wrong_comment_format",
      "go.two_fer.minimal_conditional"
    ],
    "expected_status": "disapprove_with_comment"
  },
  {
    "solution": "two-fer/solution2",
    "expected_comments": [
      "go.two_fer.stub",
      "go.two_fer.minimal_conditional"
    ],
    "not_expected_comments": [
      "go.two_fer.plus_used"
    ],
    "expected_status": "approve_with_comment"
  },
  {
    "solution": "two-fer/solution3",
    "expected_comments": [
      "go.two_fer.use_string_place_holder"
    ],
    "not_expected_comments": [],
    "expected_status": "approve_with_comment"
  }
]
```

**Comment by ErikSchierboom** [comment]
I like this idea, although it does add a little bit of complexity, but that comes with greater expressiveness. The main selling point is its extensibility, as non-maintainers should indeed be able to easily add new test cases.

One thing to consider: maybe it would make more sense to not have a single `cases.json` file per exercise, but a separate `expected_result.json` per solution? That would allow you to remove the ` "solution": "..."` part from the specification. Finding the solutions would be a simple matter of scanning for directories with an `expected_result.json` file in them. If one were to name the file `expected_result.json`, perhaps the property names themselves can also be simplified:

```json
{
    "status": "approve_with_comment",
    "comments": [
      "go.two_fer.use_string_place_holder"
    ]
}
```

Is there any particular reason to have `not_expected_comments`? Isn't it enough to verify that the comments returned exactly match the specified comments?

Overall, great idea!

**Comment by tehsphinx** [comment]
@ErikSchierboom: Good suggestions!

I like the idea to have an `expected_result.json` file per solution.

About `not_expected_comments`: If we remove that we always have to provide a complete list of expected comments. The `not_expected_comments` field helps with the following cases:
- It makes it easier for people to add a new case where one comment was missing or one comment was too much. Not having to care about the comments that were ok.
- You also might not want to have to go through all the already added solutions if you implement a new comment.

**Comment by ErikSchierboom** [comment]
> But that also forces us to update all test files if we change the output file. e.g. add a new field. I worry tests will break too easily.

I worry less about that, as I expect the output file's format to do change a bit, but not that often. And in such a case, a small search and replace should hopefully be enough.

I think that maybe we should not focus too much on standardizing, as different languages might have different requirements regarding testing. That said, I do think this is an excellent approach, and I'll be converting the C# test suite to using said approach (without the not-expected part I think ;)).

**Comment by iHiD** [comment]
Nice. https://github.com/exercism/ruby-analyzer/pull/7

**Comment by SleeplessByte** [comment]
I agree with @ErikSchierboom -- I'll implement it very similarly to the Ruby Analyzer proposal. 

Don't forget that if #11 decides we'll do templated comments, that this is more interesting as well (array of objects? Or array of ("strings" or "object")) -- and needs updating.

**Comment by tehsphinx** [comment]
I'm warming up to the idea ;)

https://github.com/exercism/go-analyzer/commit/98001ab8479046b73632720cc901c652d7214283

**Comment by SleeplessByte** [comment]
JavaScript now has this as 📸 snapshot tests:

- https://github.com/exercism/javascript-analyzer/blob/master/test/analyzers/two-fer/snapshot.ts
- https://github.com/exercism/javascript-analyzer/blob/master/test/analyzers/resistor-color/snapshot.ts

Instead of manually generating these "cases" files, this generates a snapshot of the state once, and then compares it in perpetuity. The documentation in the `javascript-analyzer` repository suggests to find ~20+ fixtures and make sure _their_ output is perfect. These are the numbers you see listed in these files. Naturally you can add all 500, but it makes sense to me to not _require_ that.
